pkg/ingester/user_tsdb.go

```diff
@@ -467,3 +499,99 @@ func (u *userTSDB) releaseAppendLock(acquireState tsdbState) {
 		u.inFlightAppendsStartedBeforeForcedCompaction.Done()
 	}
 }
+
+func (u *userTSDB) OwnedSeriesAndShards() (int, int) {
+	u.ownedSeriesMtx.Lock()
+	defer u.ownedSeriesMtx.Unlock()
+
+	return u.ownedSeriesCount, u.ownedSeriesShardSize
+}
+
+func (u *userTSDB) TriggerRecomputeOwnedSeries(reason string) {
+	u.requiresOwnedSeriesUpdate.CompareAndSwap("", reason)
+}
+
+func (u *userTSDB) RecomputeOwnedSeries(shardSize int, reason string, logger log.Logger) {
+	// We need to recompute owned series, ie. how many series in this user's Head are owned by this ingester (according to
+	// current token ranges), and updates both ownedSeries and ownedSeriesShardSize.
+
+	start := time.Now()
+	repeats := 0
+
+	ownedPrev, shardSizePrev := u.OwnedSeriesAndShards()
+
+	var ownedNew int
+	for {
+		ownedNew = u.computeOwnedSeries()
+
+		// Try to update the values, but only if no new series were added in the meantime.
+		u.ownedSeriesMtx.Lock()
+		if u.ownedSeriesCount == ownedPrev {
+			// No new series was added while computing owned series.
+			u.ownedSeriesCount = ownedNew
+			u.ownedSeriesShardSize = shardSize
+			u.ownedSeriesMtx.Unlock()
+			break
+		}
+
+		// If some series was created in the meantime, our new number of owned series may be wrong
+		// (it may or may not include the new series, we don't know). In that case, just run the computation again.
+		ownedPrev = u.ownedSeriesCount
+		shardSizePrev = u.ownedSeriesShardSize
+
+		u.ownedSeriesMtx.Unlock()
+		repeats++
+	}
+
+	dur := time.Since(start)
+
+	level.Info(logger).Log("msg", "owned series: recomputed owned series for user", "user", u.userID, "reason", reason, "ownedSeriesCountBefore", ownedPrev, "shardSizeBefore", shardSizePrev, "ownedSeriesCountAfter", ownedNew, "shardSizeAfter", shardSize, "duration", dur, "repeats", repeats)
+}
+
+// UpdateTokenRanges sets owned token ranges to supplied value, and returns true, if token ranges have changed.
+func (u *userTSDB) UpdateTokenRanges(newTokenRanges []uint32) bool {
+	// Check for changes outside critical section.
+	u.ownedTokenRangesMtx.Lock()
+	prev := u.ownedTokenRanges
+	u.ownedTokenRanges = newTokenRanges
+	u.ownedTokenRangesMtx.Unlock()
+
+	if len(newTokenRanges) != len(prev) {
+		return true
+	}
+
+	for i := 0; i < len(newTokenRanges); i++ {
+		if newTokenRanges[i] != u.ownedTokenRanges[i] {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (u *userTSDB) TokenRanges() []uint32 {
+	u.ownedTokenRangesMtx.Lock()
+	defer u.ownedTokenRangesMtx.Unlock()
+
+	return u.ownedTokenRanges
+}
+
+func (u *userTSDB) computeOwnedSeries() int {
+	u.ownedTokenRangesMtx.Lock()
```

Comment:
Do we need to compute it while holding the token range lock?

Reply:
This lock is now removed.